src/ffi/c.rs

```diff
@@ -52,14 +50,28 @@ impl Default for StreamWrapper {
                 reserved: 0,
                 opaque: ptr::null_mut(),
                 state: ptr::null_mut(),
-                #[cfg(all(feature = "any_zlib", not(feature = "cloudflare-zlib-sys")))]
-                zalloc,
-                #[cfg(all(feature = "any_zlib", not(feature = "cloudflare-zlib-sys")))]
-                zfree,
-                #[cfg(not(all(feature = "any_zlib", not(feature = "cloudflare-zlib-sys"))))]
-                zalloc: Some(zalloc),
-                #[cfg(not(all(feature = "any_zlib", not(feature = "cloudflare-zlib-sys"))))]
-                zfree: Some(zfree),
+                #[cfg(all(
+                    feature = "any_zlib",
+                    not(any(feature = "cloudflare-zlib-sys", feature = "libz-rs-sys"))
+                ))]
+                zalloc: allocator::zalloc,
+                #[cfg(all(
+                    feature = "any_zlib",
+                    not(any(feature = "cloudflare-zlib-sys", feature = "libz-rs-sys"))
+                ))]
+                zfree: allocator::zfree,
+
+                #[cfg(all(feature = "any_zlib", feature = "cloudflare-zlib-sys"))]
+                zalloc: Some(allocator::zalloc),
+                #[cfg(all(feature = "any_zlib", feature = "cloudflare-zlib-sys"))]
+                zfree: Some(allocator::zfree),
+
+                // for zlib-rs, it is most efficient to have it provide the allocator.
+                // The libz-rs-sys dependency is configured to use the rust system allocator
+                #[cfg(all(feature = "any_zlib", feature = "libz-rs-sys"))]
+                zalloc: None,
+                #[cfg(all(feature = "any_zlib", feature = "libz-rs-sys"))]
+                zfree: None,
```

Comment:
the trick is that because zlib-rs can do equality comparisons on the function pointer, it can know when its built-in allocator is used, and this allocator guarantees 64-byte alignment. The allocation of `zfree` and `zalloc` below is not needed (and that implementation is less efficient to deal with allignment).